File: ios/native/components/SNavigation2.m
//
//  SNavigation2.m
//  Supermap
//
//  Created by Asort on 2019/12/6.
//  Copyright © 2019 Facebook. All rights reserved.
//

#import "SNavigation2.h"

@implementation SNavigation2

@synthesize m_mapControl;
@synthesize m_navigation;
@synthesize m_lineDataset;
@synthesize m_pointDataset;
@synthesize m_startPoint;
@synthesize m_endPoint;
@synthesize m_nearEndPoint;
@synthesize m_nearStartPoint;
@synthesize m_modelPath;
@synthesize m_routeStyle;
@synthesize m_prjCoordSys;

-(id) initWithMapControl:(MapControl *)mapControl{
    if([super init]){
        m_mapControl = mapControl;
        m_navigation = [mapControl getNavigation2];
    }
    return self;
}
- (void) setRouteStyle:(GeoStyle *)style{
    m_routeStyle = style;
    [m_navigation setRouteStyle:style];
}
-(void) setNetworkDataset:(DatasetVector *)dataset{
    m_lineDataset = dataset;
    m_pointDataset = dataset.childDataset;
    [m_navigation setNetworkDataset:dataset];
}

-(void) setStartPoint:(double)x sPointY:(double)y{
    m_startPoint = [[Point2D alloc] initWithX:x Y:y];
    [m_navigation setStartPoint:x sPointY:y];
}

-(void) setDestinationPoint:(double)x dPointY:(double)y{
    m_endPoint = [[Point2D alloc] initWithX:x Y:y];
    [m_navigation setDestinationPoint:x dPointY:y];
}

- (BOOL) loadModel:(NSString *)filePath{
    m_modelPath = filePath;
    BOOL isLoaded = [m_navigation loadModel:filePath];
    return isLoaded;
}
- (BOOL) reAnalyst{
    @try{
    GeoPoint *sPoint = [[GeoPoint alloc] initWithPoint2D:m_startPoint];
    DatasetVector *datasetVector = (DatasetVector *)m_pointDataset;
    Recordset *rd = [datasetVector queryWithGeometry:sPoint BufferDistance:0.001 Type:STATIC];
    if([rd recordCount] == 0){
        rd = [datasetVector queryWithGeometry:sPoint BufferDistance:0.005 Type:STATIC];
    }
    Geometry *geo  = [[Geometry alloc] init];
    double dLen = 1000000.0;
    while (![rd isEOF]) {
        geo = rd.geometry;
        Point2D *tmpPoint = [geo getInnerPoint];
        double len = sqrt( (m_startPoint.x - tmpPoint.x)*(m_startPoint.x - tmpPoint.x) + (m_startPoint.y - tmpPoint.y)*(m_startPoint.y - tmpPoint.y) );
        if(dLen > len){
            Recordset *lineRd = [m_lineDataset queryWithGeometry:geo BufferDistance:0 Type:STATIC];
            if([lineRd recordCount] != 0){
                dLen = len;
                m_nearStartPoint = tmpPoint;
            }
            [lineRd close];
            [lineRd dispose];
        }
        [rd moveNext];
    }
    [rd close];
    [rd dispose];
    
    GeoPoint *ePoint = [[GeoPoint alloc] initWithPoint2D:m_endPoint];
    Recordset *rd2 = [datasetVector queryWithGeometry:ePoint BufferDistance:0.001 Type:STATIC];
    if([rd2 recordCount] == 0){
        rd2 = [datasetVector queryWithGeometry:ePoint BufferDistance:0.005 Type:STATIC];
    }
    double dLen2 = 1000000.0;
    while (![rd2 isEOF]) {
        geo = rd2.geometry;
        Point2D *tmpPoint = [geo getInnerPoint];
        double len = sqrt( (m_endPoint.x - tmpPoint.x)*(m_endPoint.x - tmpPoint.x) + (m_endPoint.y - tmpPoint.y)*(m_endPoint.y - tmpPoint.y) );
        if(dLen2 > len){
            Recordset *lineRd = [m_lineDataset queryWithGeometry:geo BufferDistance:0 Type:STATIC];
            if([lineRd recordCount] != 0){
                dLen2 = len;
                m_nearEndPoint = tmpPoint;
            }
            [lineRd close];
            [lineRd dispose];
        }
        [rd2 moveNext];
    }
    [rd2 close];
    [rd2 dispose];
    [geo dispose];
    
    BOOL isFind = NO;
    if(m_nearStartPoint != nil && m_nearEndPoint != nil){
        [m_navigation setStartPoint:m_nearStartPoint.x sPointY:m_nearStartPoint.y];
        [m_navigation setDestinationPoint:m_nearEndPoint.x dPointY:m_nearEndPoint.y];
        isFind = [m_navigation routeAnalyst];
        if(!isFind){
            m_startPoint = nil;
            m_endPoint = nil;
            m_nearStartPoint = nil;
            m_nearEndPoint = nil;
        }
    }
    
    return isFind;
    }@catch(NSException *e){
        NSLog(@"%@",e.reason);
    }
}
-(void) addGuideLineOnTrackinglayerWithMapPrj:(PrjCoordSys *)mapPrjCoordSys{
    m_prjCoordSys = mapPrjCoordSys;
    
    Point2D *start = [SNavigation2 getMapPointWithPoint:m_startPoint PrjCoordSys:mapPrjCoordSys];
    Point2D *nearStart = [SNavigation2 getMapPointWithPoint:m_nearStartPoint PrjCoordSys:mapPrjCoordSys];
    Point2D *end = [SNavigation2 getMapPointWithPoint:m_endPoint PrjCoordSys:mapPrjCoordSys];
    Point2D *nearEnd = [SNavigation2 getMapPointWithPoint:m_nearEndPoint PrjCoordSys:mapPrjCoordSys];
    
    TrackingLayer *layer = m_mapControl.map.trackingLayer;
        
    GeoStyle *guideLine = [[GeoStyle alloc] init];
    [guideLine setLineWidth:2];
    [guideLine setLineColor:[[Color alloc]initWithR:82 G:198 B:233]];
    [guideLine setLineSymbolID:2];
        
    Point2Ds *startPoints = [[Point2Ds alloc] init];
    [startPoints add:start];
    [startPoints add:nearStart];
        
    GeoLine *startLine = [[GeoLine alloc] initWithPoint2Ds:startPoints];
    [startLine setStyle:guideLine];
    [layer addGeometry:startLine WithTag:@"startLine"];
        
    Point2Ds *endPoints = [[Point2Ds alloc] init];
    [endPoints add:end];
    [endPoints add:nearEnd];
        
    GeoLine *endLine = [[GeoLine alloc] initWithPoint2Ds:endPoints];
    [endLine setStyle:guideLine];
    [layer addGeometry:endLine WithTag:@"endLine"];
        
    [m_mapControl.map refresh];
    
    m_startPoint = nil;
    m_endPoint = nil;
    m_nearStartPoint = nil;
    m_nearEndPoint = nil;
}

+(Point2D *)getMapPointWithPoint:(Point2D *)pt PrjCoordSys:(PrjCoordSys *) prjCoordSys{
    Point2D *point2D = nil;
    if(pt.x >= -180 && pt.x <= 180 && pt.y >= -90 && pt.y <= 90){
        Point2Ds *points = [[Point2Ds alloc]init];
        [points add:pt];
        PrjCoordSys *desPrjCoordSys = [[PrjCoordSys alloc]init];
        desPrjCoordSys.type = PCST_EARTH_LONGITUDE_LATITUDE;
        [CoordSysTranslator convert:points PrjCoordSys:desPrjCoordSys PrjCoordSys:prjCoordSys CoordSysTransParameter:[[CoordSysTransParameter alloc]init] CoordSysTransMethod:MTH_GEOCENTRIC_TRANSLATION];
        point2D = [points getItem:0];
    }else{
        point2D = pt;
    }
    return point2D;
}

@end

File: android/src/main/java/com/supermap/smNative/components/SNavigation2.java
package com.supermap.smNative.components;

import android.util.Log;

import com.supermap.data.Color;
import com.supermap.data.CoordSysTransMethod;
import com.supermap.data.CoordSysTransParameter;
import com.supermap.data.CoordSysTranslator;
import com.supermap.data.CursorType;
import com.supermap.data.Dataset;
import com.supermap.data.DatasetVector;
import com.supermap.data.GeoLine;
import com.supermap.data.GeoPoint;
import com.supermap.data.GeoStyle;
import com.supermap.data.Geometry;
import com.supermap.data.Point2D;
import com.supermap.data.Point2Ds;
import com.supermap.data.PrjCoordSys;
import com.supermap.data.PrjCoordSysType;
import com.supermap.data.Recordset;
import com.supermap.mapping.MapControl;
import com.supermap.mapping.TrackingLayer;
import com.supermap.navi.Navigation2;

/**
 * Author:     Asort
 * Date:        2019/12/9
 * project:     iTablet
 * package:     iTablet
 * class:
 * description:
 */

public class SNavigation2 {
    private MapControl mapControl;
    private Navigation2 navigation;
    private Point2D startPoint;
    private Point2D endPoint;
    private Point2D nearStartPoint;
    private Point2D nearEndPoint;
    private DatasetVector lineDataset;
    private Dataset pointDataset;
    private String modelPath;
    private GeoStyle routeStyle;
    private PrjCoordSys prjCoordSys;

    public Navigation2 getNavigation(){
        return navigation;
    }
    public SNavigation2(MapControl mapController){
        mapControl = mapController;
        navigation = mapControl.getNavigation2();
    }

    public void setNetworkDataset(DatasetVector dataset){
        lineDataset = dataset;
        pointDataset = dataset.getChildDataset();
        navigation.setNetworkDataset(dataset);
    }

    public void setRouteStyle(GeoStyle style){
        routeStyle = style;
        navigation.setRouteStyle(style);
    }

    public void setStartPoint(double x, double y) {
        startPoint = new Point2D(x, y);
        navigation.setStartPoint(x,y);
    }

    public void setDestinationPoint(double x, double y) {
        endPoint = new Point2D(x, y);
        navigation.setDestinationPoint(x,y);
    }

    public boolean loadModel(String filePath){
        modelPath = filePath;
        return navigation.loadModel(filePath);
    }

    /**
     * 重新进行路径分析 保证分析结果存在
     * @return boolean
     */
    public boolean reAnalyst(){
        try{
            GeoPoint sPoint = new GeoPoint(startPoint);
            DatasetVector datasetVector = (DatasetVector)pointDataset;
            Recordset recordset = datasetVector.query(sPoint,0.001,CursorType.STATIC);
            if(recordset.getRecordCount() == 0){
                recordset = datasetVector.query(sPoint,0.005,CursorType.STATIC);
            }
            Geometry geo = null;
            double dLen = 1000000.0;
            while (!recordset.isEOF()){
                geo = recordset.getGeometry();
                Point2D tmpPoint = geo.getInnerPoint();
                double len = Math.sqrt( Math.pow( (startPoint.getX() - tmpPoint.getX()),2) + Math.pow( (startPoint.getY() - tmpPoint.getY()),2));
                if(dLen > len){
                    Recordset lineRecordset = lineDataset.query(geo,0,CursorType.STATIC);
                    if(lineRecordset.getRecordCount() != 0){
                        dLen = len;
                        nearStartPoint = tmpPoint;
                    }
                    lineRecordset.close();
                    lineRecordset.dispose();
                }

                recordset.moveNext();
            }
            recordset.close();
            recordset.dispose();

            GeoPoint ePoint = new GeoPoint(endPoint);
            Recordset recordset1 = datasetVector.query(ePoint,0.001,CursorType.STATIC);
            if(recordset1.getRecordCount() == 0){
                recordset1 = datasetVector.query(ePoint,0.005,CursorType.STATIC);
            }

            double dLen1 = 1000000.0;
            while (!recordset1.isEOF()){
                geo = recordset1.getGeometry();
                Point2D tmpPoint = geo.getInnerPoint();
                double len = Math.sqrt( Math.pow( (endPoint.getX() - tmpPoint.getX()),2) + Math.pow( (endPoint.getY() - tmpPoint.getY()),2));
                if(dLen1 > len){
                    Recordset lineRecordset = lineDataset.query(geo,0,CursorType.STATIC);
                    if(lineRecordset.getRecordCount() != 0){
                        dLen1 = len;
                        nearEndPoint = tmpPoint;
                    }
                    lineRecordset.close();
                    lineRecordset.dispose();
                }

                recordset1.moveNext();
            }
            recordset1.close();
            recordset1.dispose();
            if(geo != null){
                geo.dispose();
            }

            boolean isFind = false;
            if(nearStartPoint != null && nearEndPoint != null){
                navigation.setStartPoint(nearStartPoint.getX(),nearStartPoint.getY());
                navigation.setDestinationPoint(nearEndPoint.getX(),nearEndPoint.getY());
                isFind = navigation.routeAnalyst();
                if(!isFind){
                    startPoint = null;
                    endPoint = null;
                    nearStartPoint = null;
                    nearEndPoint = null;
                }
            }
            return isFind;
        }catch (Exception e){
            Log.e("ReAnalyst Error",e.toString());
            return false;
        }
    }

    /**
     * 添加引导线（起点和起点邻近点、终点和终点临界点的 之间的虚线）
     * @param mapPrjCoordSys
     */
    public void addGuideLineOnTrackingLayer(PrjCoordSys mapPrjCoordSys){
        prjCoordSys = mapPrjCoordSys;
        Point2D start = getMapPoint(startPoint,prjCoordSys);
        Point2D nearStart = getMapPoint(nearStartPoint,prjCoordSys);
        Point2D end = getMapPoint(endPoint,prjCoordSys);
        Point2D nearEnd = getMapPoint(nearEndPoint,prjCoordSys);

        TrackingLayer layer = mapControl.getMap().getTrackingLayer();

        GeoStyle guideLine = new GeoStyle();
        guideLine.setLineWidth(2.0);
        guideLine.setLineColor(new Color(82,198,233));
        guideLine.setLineSymbolID(2);

        Point2Ds startPoints = new Point2Ds();
        startPoints.add(start);
        startPoints.add(nearStart);

        GeoLine startLine = new GeoLine(startPoints);
        startLine.setStyle(guideLine);
        layer.add(startLine,"startLine");

        Point2Ds endPoints = new Point2Ds();
        endPoints.add(end);
        endPoints.add(nearEnd);

        GeoLine endLine = new GeoLine(endPoints);
        endLine.setStyle(routeStyle);
        layer.add(endLine,"endLine");

        mapControl.getMap().refresh();

        startPoint = null;
        endPoint = null;
        nearStartPoint = null;
        nearEndPoint = null;
    }

    private Point2D getMapPoint(Point2D pt, PrjCoordSys prjCoordSys){
        Point2D point2D = null;
        double x = pt.getX();
        double y = pt.getY();
        if(x > -180 && x < 180 && y > -90 && y < 90){
            Point2Ds point2Ds = new Point2Ds();
            point2Ds.add(pt);
            PrjCoordSys desPrjCoordSys = new PrjCoordSys();
            desPrjCoordSys.setType(PrjCoordSysType.PCS_EARTH_LONGITUDE_LATITUDE);
            CoordSysTranslator.convert(point2Ds,desPrjCoordSys,prjCoordSys,new CoordSysTransParameter(),CoordSysTransMethod.MTH_GEOCENTRIC_TRANSLATION);
            point2D = point2Ds.getItem(0);
        }else{
            point2D = pt;
        }
        return point2D;
    }
}
